src/Microsoft.AspNet.Routing/InlineRouteParameterParser.cs

```diff
@@ -12,82 +9,230 @@ namespace Microsoft.AspNet.Routing
+            if (routeParameter[endIndex] == '?')
+            {
+                isOptional = true;
+                endIndex--;
+            }
+
+            var currentIndex = startIndex;
 
-            routeParameter = isCatchAll ? routeParameter.Substring(1) : routeParameter;
-            routeParameter = isOptional ? routeParameter.Substring(0, routeParameter.Length - 1) : routeParameter;
+            // Parse parameter name
+            var parameterName = string.Empty;
 
-            var parameterMatch = _parameterRegex.Match(routeParameter);
-            if (!parameterMatch.Success)
+            while (currentIndex <= endIndex)
             {
-                return TemplatePart.CreateParameter(name: string.Empty,
-                                                    isCatchAll: isCatchAll,
-                                                    isOptional: isOptional,
-                                                    defaultValue: null,
-                                                    inlineConstraints: null);
+                var currentChar = routeParameter[currentIndex];
+
+                if (currentIndex == endIndex)
+                {
+                    parameterName = routeParameter.Substring(startIndex, currentIndex - startIndex + 1);
+                }
+                else if ((currentChar == ':' || currentChar == '=') && startIndex != currentIndex)
+                {
+                    // Parameter names are allowed to start with delimiters used to denote constraints or default values.
+                    // i.e. "=foo" or ":bar" would be treated as parameter names rather than default value or constraint
+                    // specifications.
+                    parameterName = routeParameter.Substring(startIndex, currentIndex - startIndex);
+
+                    // Roll the index back and move to the constraint parsing stage.
+                    currentIndex--;
+                    break;
+                }
+
+                currentIndex++;
             }
 
-            var parameterName = parameterMatch.Groups["parameterName"].Value;
+            var parseResults = ParseConstraints(routeParameter, currentIndex, endIndex);
+            currentIndex = parseResults.CurrentIndex;
 
-            // Add the default value if present
-            var defaultValueGroup = parameterMatch.Groups["defaultValue"];
-            var defaultValue = GetDefaultValue(defaultValueGroup);
-
-            // Register inline constraints if present
-            var constraintGroup = parameterMatch.Groups["constraint"];
-            var inlineConstraints = GetInlineConstraints(constraintGroup);
+            string defaultValue = null;
+            if (currentIndex <= endIndex &&
+                routeParameter[currentIndex] == '=')
+            {
+                defaultValue = routeParameter.Substring(currentIndex + 1, endIndex - currentIndex);
+            }
```

Comment:
Although this did uncover a bug in the current implementation. Adding tests for it.